toolchain/docs/check/pattern_matching.md

```diff
@@ -0,0 +1,285 @@
+advance: any given subpattern might turn out to be one that emits procedural
+insts. To handle these situations, we speculatively push an inst block onto the
+(procedural) stack whenever we are about to begin handling a subpattern, and
+then pop it at the end of the subpattern, with different treatment depending on
+whether the subpattern turned out to be a subexpression. This is handled by
+`BeginSubpattern`, `EndSubpatternAsExpr`, and `EndSubpatternAsNonExpr`.
+
+One further complication here is that the type expression can contain control
+flow (such as an `if` expression). Consequently, we can't represent the type
+expression SemIR as a single block; instead, we represent the SemIR for a given
+type expression as a
+[single-entry, single-exit (SE/SE) region](https://en.wikipedia.org/wiki/Single-entry_single-exit),
+potentially consisting of multiple blocks.
+
+> **Note:** The original motivation for rigorously excluding procedural insts
+> from the pattern block may no longer apply. In particular, it may make sense
+> to put procedural insts in the pattern block when they represent an expression
+> that is part of the pattern. If so, substantial parts of this design might
+> change. See
+> [issue #5351](https://github.com/carbon-language/carbon-lang/issues/5351).
+
+## Parser-driven pattern block pushing
+
+At the same time as all of that, we have to manage the _pattern_ block stack as
+well. We attempt to do this precisely rather than speculatively, by leveraging
+the parser to precisely mark the nodes immediately before full-patterns, and
+pushing the pattern block stack when we handle those nodes. We then rely on
+signals from both the parser and the node stack to determine when to pop from
+the pattern block stack.
+
+In the case of `let` and `var` decls, this is fairly straightforward: the
+beginning is marked by the `LetIntroducer` or `VarIntroducer` node, and the end
+is marked by the `LetInitializer` or `VarInitializer`, or by the `VarDecl` in
+the case of a `var` decl with no initializer. Similarly, the beginning of an
+`impl forall` parameter list is marked by the `Forall` node, and the end is
+marked by the `ImplDecl` or `ImplDefinitionStart`.
+
+The case of a parameterized name (such as `Foo(x: i32)`) is more challenging.
+The node immediately before the start of the full-pattern is an identifier, but
+an identifier doesn't necessarily mark the start of a full-pattern. We've solved
+that by having the parser mark identifier nodes that are followed by
+full-patterns (using lookahead). Rather than use additional storage for what is
+logically a single bit of data, we effectively smuggle that bit into the kind
+enum by having separate node kinds `IdentifierNameBeforeParams` and
+`IdentifierNameNotBeforeParams`.
+
+If the parameterized name is a name qualifier (such as the first part of
+`Foo(X: i32).Bar(y: i32)`), the node immediately after it will be the qualifier
```

Comment:
Typo?

```suggestion
`Foo(x: i32).Bar(y: i32)`), the node immediately after it will be the qualifier
```

Reply:
Yes, but I think the typo was omitting the `!`, not the capitalization.